### PR TITLE
Ensure pending transport handlers are invoked for all channel failures

### PR DESCRIPTION
**Pull request by s1monw**
Today if a channel gets closed due to a disconnect we notify the response
handler that the connection is closed and the node is disconnected. Unfortunately
this is not a complete solution since it only works for published connections.
Connections that are unpublished ie. for discovery can indefinitely hang since we
never invoke their handers when we get a failure while a user is waiting for
the response. This change adds connection tracking to TcpTransport that ensures
we are notifying the corresponding connection if there is a failure on a channel.

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine test this please

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 1]
```diff
@@ -580,17 +585,27 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
...
-                    for (DiscoveryNode node : connectedNodes.keySet()) {
-                        if (disconnectFromNode(node, channel, ExceptionsHelper.detailedMessage(failure))) {
+                    for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
+                        if (disconnectFromNode(entry.getKey(), channel, reason)) {
                             // if we managed to find this channel and disconnect from it, then break, no need to check on
                             // the rest of the nodes
-                            break;
+                            // Removing it here from open connections is paranoia since the close handler should deal with this.
+                            openConnections.remove(entry.getValue());
+                            return;
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Are you sure you want this? Won't this cause any exception thrown by `closeChannels` to be lost?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 2]
```diff
@@ -580,17 +585,27 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
...
+                        }
+                    }
+                    // now if we haven't found the right connection in the connected nodes we have to go through the open connections
+                    // it might be that the channel belongs to a connection that is not published
+                    for (NodeChannels channels : openConnections) {
+                        if (channels.hasChannel(channel)) {
+                            IOUtils.closeWhileHandlingException(channels);
+                            return;
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Same comment

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: And there is still a catch block there to log that exception, so I assume we still want it?

Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Also there is like double try blocks that I'm not sure need to be nested.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 3]
```diff
@@ -902,11 +917,8 @@ protected final void doStop() {
                     }
                 }
 
-                for (Iterator<NodeChannels> it = connectedNodes.values().iterator(); it.hasNext();) {
-                    NodeChannels nodeChannels = it.next();
-                    it.remove();
-                    IOUtils.closeWhileHandlingException(nodeChannels);
-                }
+                IOUtils.closeWhileHandlingException(() -> IOUtils.closeWhileHandlingException(connectedNodes.values()),
+                    () -> IOUtils.closeWhileHandlingException(openConnections), openConnections::clear, connectedNodes::clear);
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: It is reasonably challenging to figure out what is going on here. Maybe like a comment about the order of operations or something here?

I think it's closing nodes. Followed by closing any "light" channels. And then clearing the maps?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java [hunk 4]
```diff
@@ -374,21 +374,17 @@ protected NodeChannels connectToChannels(DiscoveryNode node, ConnectionProfile p
...
-
+    private final class ChannelCloseListener implements ChannelFutureListener {
         @Override
         public void operationComplete(final ChannelFuture future) throws Exception {
-            onChannelClosed(future.channel());
-            NodeChannels nodeChannels = connectedNodes.get(node);
-            if (nodeChannels != null && nodeChannels.hasChannel(future.channel())) {
-                threadPool.generic().execute(() -> disconnectFromNode(node, future.channel(), "channel closed event"));
+            try {
+                onChannelClosed(future.channel());
```
Comment (tbrooks8) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: I know there are are a bunch of checks at various points so that we can call this close methods multiple times safely. But I just want to point out that in this case:

1. We call `onChannelClosed` here
2. In the finally block below we call `disconnectFromNodeChannel` which dispatches a Runnable to a generic thread pool.
3. At the end of `disconnectFromNodeChannel` we call `onChannelClosed` again with the same channel.

That is all probably fine. But this all seems to be pretty complicated to me right now. Maybe we can find some ways to simplify in the future?

**Review by Tim-Brooks — CHANGES_REQUESTED** [review]
I made some comments.

**Comment by tbrooks8** [comment]
I do want to clarify, for my understanding, the core of this PR is here correct?

```
for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
                        if (disconnectFromNode(entry.getKey(), channel, reason)) {
                            // if we managed to find this channel and disconnect from it, then break, no need to check on
                            // the rest of the nodes
                            // Removing it here from open connections is paranoia since the close handler should deal with this.
                            openConnections.remove(entry.getValue());
                            return;
                        }
                    }
                    // now if we haven't found the right connection in the connected nodes we have to go through the open connections
                    // it might be that the channel belongs to a connection that is not published
                    for (NodeChannels channels : openConnections) {
                        if (channels.hasChannel(channel)) {
                            IOUtils.closeWhileHandlingException(channels);
                            return;
                        }
                    }
```

This is where you're checking, if this is not a fully connected node, check ALL the connections and close if it is found?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java [hunk 5]
```diff
@@ -345,7 +345,7 @@ protected NodeChannels connectToChannels(DiscoveryNode node, ConnectionProfile p
                         throw new ConnectTransportException(node, "connect_timeout[" + connectTimeout + "]", future.cause());
                     }
                     channels[i] = future.channel();
-                    channels[i].closeFuture().addListener(new ChannelCloseListener(node));
+                    channels[i].closeFuture().addListener(new ChannelCloseListener());
```
Comment (s1monw) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: ++

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java (hunk 4 above)
Comment (s1monw) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: you are right, I simplified this by simply calling `disconnectFromNodeChannel` from `onChannelClosed` only. It's much simpler now but I think we can improve it further down the road.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I simplified this and added a comment. the double try is likely a leftover..

**Comment by s1monw** [comment]
> This is where you're checking, if this is not a fully connected node, check ALL the connections and close if it is found?

correct!

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 3 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I simplified it in a new commit

**Comment by s1monw** [comment]
great feedback @tbrooks8  I pushed a new commit to address all your comments.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 6]
```diff
@@ -580,24 +587,37 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
     /**
      * Disconnects from a node if a channel is found as part of that nodes channels.
      */
-    protected final void disconnectFromNodeChannel(final Channel channel, final Exception failure) {
+    protected final void disconnectFromNodeChannel(final Channel channel, final String reason) {
         threadPool.generic().execute(() -> {
             try {
-                try {
+                if (isOpen(channel)) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: question - why did you add this? it seems all channels has double closing protection?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 7]
```diff
@@ -580,24 +587,37 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
...
+                logger.warn("failed to close channel", e);
+            } finally {
+                outer:
+                {
+                    for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
+                        if (disconnectFromNode(entry.getKey(), channel, reason)) {
                             // if we managed to find this channel and disconnect from it, then break, no need to check on
                             // the rest of the nodes
+                            // Removing it here from open connections is paranoia since the close handler should deal with this.
+                            openConnections.remove(entry.getValue());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: given the comment, should we make this an assertion?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 8]
```diff
@@ -901,12 +921,11 @@ protected final void doStop() {
                                 "Error closing serverChannel for profile [{}]", entry.getKey()), e);
                     }
                 }
-
-                for (Iterator<NodeChannels> it = connectedNodes.values().iterator(); it.hasNext();) {
-                    NodeChannels nodeChannels = it.next();
-                    it.remove();
-                    IOUtils.closeWhileHandlingException(nodeChannels);
-                }
+                // we are holding a write lock so nobody modifies the connectedNodes / openConnections map - it's safe to first close
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: ++

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 9]
```diff
@@ -580,24 +587,37 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
...
+                            // Removing it here from open connections is paranoia since the close handler should deal with this.
+                            openConnections.remove(entry.getValue());
+                            // we can only be connected and published to a single node with one connection. So if disconnectFromNode
+                            // returns true we can safely break out from here since we cleaned up everything needed
+                            break outer;
+                        }
+                    }
+                    // now if we haven't found the right connection in the connected nodes we have to go through the open connections
+                    // it might be that the channel belongs to a connection that is not published
+                    for (NodeChannels channels : openConnections) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I think we need some kind of locking coordination here with the opening connections line:

```
                nodeChannels = connectToChannels(node, connectionProfile); <-- from here we might get exceptions
                final Channel channel = nodeChannels.getChannels().get(0); // one channel is guaranteed by the connection profile
                final TimeValue connectTimeout = connectionProfile.getConnectTimeout() == null ?
                    defaultConnectionProfile.getConnectTimeout() :
                    connectionProfile.getConnectTimeout();
                final TimeValue handshakeTimeout = connectionProfile.getHandshakeTimeout() == null ?
                    connectTimeout : connectionProfile.getHandshakeTimeout();
                final Version version = executeHandshake(node, channel, handshakeTimeout);
                nodeChannels = new NodeChannels(nodeChannels, version); // clone the channels - we now have the correct version
                openConnections.add(nodeChannels); <-- published 
                transportServiceAdapter.onConnectionOpened(nodeChannels);
```

I guess the easiest would be to acquire the write lock as this is a very rare situation and only relevant to temporary connections. Alternatively we would could introduce a dedicated lock.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 10]
```diff
@@ -1608,16 +1629,20 @@ public long newRequestId() {
...
+                final Long requestId = first.get();
+                final HandshakeResponseHandler handler = pendingHandshakes.remove(requestId);
+                if (handler != null) {
+                    // there might be a race removing this or this method might be called twice concurrently depending on how
+                    // the channel is closed ie. due to connection reset or broken pipes
+                    handler.handleException(new TransportException("connection reset"));
+                }
             }
+        } finally {
+            disconnectFromNodeChannel(channel, "channel closed");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: question - why did you add this? it currently has a nasty side effect of spawning extra "disconnect" generic threads when we are already disconnecting. Say we have an issue on one channel, we find the relevant NodeChannels and close all other channels in it. Those will end up here, spawning another disconnectFromNodeChannel call, which will be a no-op (as we are cleaning it up), but we're creating all kind of concurrency issues. I'm wondering why we need this.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java [hunk 11]
```diff
@@ -374,24 +375,6 @@ protected NodeChannels connectToChannels(DiscoveryNode node, ConnectionProfile p
         return nodeChannels;
     }
 
-    private class ChannelCloseListener implements ChannelFutureListener {
```
Comment (bleskes) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: ++ to push this up to TcpTransport

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 12]
```diff
@@ -190,6 +194,13 @@ public void assertNoPendingHandshakes(Transport transport) {
         }
     }
 
+    public void assertPendingConnections(int numConnections, Transport transport) {
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: ❤️

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 13]
```diff
@@ -1852,7 +1861,11 @@ public String executor() {
         }
         logger.debug("DONE");
         serviceC.close();
-
+        // when we close C here we have to disconnect the service otherwise assertions mit trip with pending connections in tearDown
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: type: mit -> might

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 14]
```diff
@@ -2155,12 +2173,83 @@ public String executor() {
...
+
+        try (Transport.Connection connection = serviceB.openConnection(serviceC.getLocalNode(), builder.build())) {
+            serviceB.sendRequest(connection, "action", new TestRequest("hello world"), TransportRequestOptions.EMPTY,
+                transportResponseHandler);
+            receivedLatch.await();
+            assertPendingConnections(1, serviceB.getOriginalTransport());
+            serviceC.close();
+            assertPendingConnections(0, serviceC.getOriginalTransport());
+            sendResponseLatch.countDown();
+            responseLatch.await();
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: I would expect the request to be cancelled as the connection closed on it mid flight. What is the role of sendResponseLatch? serviceC is close now?

**Review by bleskes — CHANGES_REQUESTED** [review]
This is a great catch and the solution is nice too. I left a few small questions/comments on concurrency aspects.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 10 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: you need this since otherwise we might not notify a the listeners waiting for an response on the connection. We might be able to reduce it's footprint more. I can look but it's already improved and preexsting hence why I don't wanted to touch it more. The call was hidden by other calls that's why you don't see it

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 9 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: why do we need this, this is only relevant for fully established connections.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 6 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: I left a comment in the code for this

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java (hunk 14 above)
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: if I don't keep this latch the test is useless since we are not stopping in between. I can remove the countDown here and that would still work but it feels wrong. Let it deal with what happens if you try to respond to a closed channel?

Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: it's also relevant for some implemenations that we let things go threadwise if there is only one event thread. they will block.

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java (hunk 9 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: @s1monw and I discussed this and apparently transportServiceAdapter spawns threads and any point of synchronization is moot and only a source of trouble (blocking on netty is super tricky)

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java (hunk 14 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: We talked this one too - the extra `sendResponseLatch.countDown();` is needed to release a netty thread so that netty won't be annoyed. We agreed to add a comment.

**Review by bleskes — APPROVED** [review]

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Comment by s1monw** [comment]
@elasticmachine test this please
